src/server/pfs/s3/multipart.go

```diff
@@ -0,0 +1,412 @@
+package s3
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/gorilla/mux"
+	"github.com/pachyderm/pachyderm/src/client"
+	pfsClient "github.com/pachyderm/pachyderm/src/client/pfs"
+	pfsServer "github.com/pachyderm/pachyderm/src/server/pfs"
+	"github.com/pachyderm/pachyderm/src/server/pkg/errutil"
+	"github.com/pachyderm/pachyderm/src/server/pkg/uuid"
+	"github.com/pachyderm/s2"
+)
+
+var multipartChunkPathMatcher = regexp.MustCompile(`([^/]+)/([^/]+)/(.+)/([^/]+)/(\d+)`)
+var multipartKeepPathMatcher = regexp.MustCompile(`([^/]+)/([^/]+)/(.+)/([^/]+)/\.keep`)
+
+func multipartChunkArgs(path string) (repo string, branch string, key string, uploadID string, partNumber int, err error) {
+	match := multipartChunkPathMatcher.FindStringSubmatch(path)
+
+	if len(match) == 0 {
+		err = errors.New("invalid file path found in multipath bucket")
+		return
+	}
+
+	repo = match[1]
+	branch = match[2]
+	key = match[3]
+	uploadID = match[4]
+	partNumber, err = strconv.Atoi(match[5])
+	if err != nil {
+		err = fmt.Errorf("invalid file path found in multipath bucket: %s", err)
+		return
+	}
+	return
+}
+
+func multipartKeepArgs(path string) (repo string, branch string, key string, uploadID string, err error) {
+	match := multipartKeepPathMatcher.FindStringSubmatch(path)
+
+	if len(match) == 0 {
+		err = errors.New("invalid file path found in multipath bucket")
+		return
+	}
+
+	repo = match[1]
+	branch = match[2]
+	key = match[3]
+	uploadID = match[4]
+	return
+}
+
+func parentDirPath(repo, branch, key, uploadID string) string {
+	return fmt.Sprintf("%s/%s/%s/%s", repo, branch, key, uploadID)
+}
+
+func chunkPath(repo, branch, key, uploadID string, partNumber int) string {
+	return fmt.Sprintf("%s/%d", parentDirPath(repo, branch, key, uploadID), partNumber)
+}
+
+func keepPath(repo, branch, key, uploadID string) string {
+	return fmt.Sprintf("%s/.keep", parentDirPath(repo, branch, key, uploadID))
+}
+
+func (c *controller) ensureRepo(pc *client.APIClient) error {
+	_, err := pc.InspectBranch(c.repo, "master")
+	if err != nil {
+		err = pc.CreateRepo(c.repo)
+		if err != nil && !strings.Contains(err.Error(), "as it already exists") {
+			return err
+		}
+
+		err = pc.CreateBranch(c.repo, "master", "", nil)
+		if err != nil && !strings.Contains(err.Error(), "as it already exists") {
+			return err
+		}
```

Comment:
Similarly, I think you can drop this error filter since `CreateBranch` allows you to update branches by default since that's such a common operation.